nomos-blend/message/src/reward.rs

```diff
@@ -0,0 +1,521 @@
+            ),
+        ));
+        self.current_session_randomness = new_session_info.session_randomness;
+    }
+}
+
+/// Arguments required to build [`BlendingTokens`].
+pub struct SessionInfo {
+    session_number: SessionNumber,
+    session_randomness: SessionRandomness,
+    num_core_nodes: u64,
+    total_core_quota: u64,
+    message_frequency_per_round: NonNegativeF64,
+}
+
+/// Blending tokens collected during a single session.
+struct SessionBlendingTokens {
+    session_number: SessionNumber,
+    num_core_nodes: u64,
+    total_core_quota: u64,
+    message_frequency_per_round: NonNegativeF64,
+    tokens: HashSet<BlendingToken>,
+}
+
+impl SessionBlendingTokens {
+    fn new(
+        session_number: SessionNumber,
+        num_core_nodes: u64,
+        total_core_quota: u64,
+        message_frequency_per_round: NonNegativeF64,
+    ) -> Self {
+        Self {
+            session_number,
+            num_core_nodes,
+            total_core_quota,
+            message_frequency_per_round,
+            tokens: HashSet::new(),
+        }
+    }
+
+    fn insert(&mut self, token: BlendingToken) {
+        self.tokens.insert(token);
+    }
+
+    /// Computes an activity proof for this session by consuming tokens.
+    ///
+    /// It returns `Ok(None)` if there was no blending token satisfying the
+    /// activity threshold calculated.
+    /// It returns `Err` if the computation fails.
+    fn activity_proof(
+        self,
+        next_session_randomness: SessionRandomness,
+    ) -> Result<Option<ActivityProof>, Error> {
+        // First, compute the activity threshold.
+        let activity_threshold = activity_threshold(
+            self.num_core_nodes,
+            BlendingToken::repr_bit_len(self.total_core_quota, self.message_frequency_per_round)?,
+        )?;
+        debug!(
+            "Activity threshold for session {}: {}",
+            self.session_number, activity_threshold
+        );
+
+        // Filter tokens whose Hamming distance is below the activity threshold,\
+        // and pick the one with the smallest distance.
+        let maybe_token = self
+            .tokens
+            .into_iter()
+            .filter_map(|token| {
+                match token.hamming_distance(
+                    self.message_frequency_per_round,
+                    self.total_core_quota,
+                    next_session_randomness,
+                ) {
+                    Ok(distance) => {
+                        debug!("Token distance:{distance}, threshold: {activity_threshold}");
+                        (distance <= activity_threshold).then_some((token, distance))
+                    }
+                    Err(e) => {
+                        error!("Failed to compute Hamming distance: {e}");
+                        None
+                    }
+                }
+            })
+            .min_by_key(|(_, distance)| *distance);
+        let Some((token, _)) = maybe_token else {
```

Comment:
If we fail on construction instead, this function could then finally just return an `Option` instead of a `Result<Option>`, which I think it's cleaner.